### PR TITLE
feat(proposals/2021): add information for repo structure

### PR DESCRIPTION
**Pull request by jbrockopp**
This proposes to condense the current repo structure in the [go-vela](https://github.com/go-vela) org.

Currently, we have `11` "core" repos for Vela.

After this proposal, we'd be left with `6` "core" repos for Vela.

> NOTE:
>
> In this context, a "core" repo is any repos that require code or dependency changes in order to release the product.

The intention of this proposal is to accomplish the following:

* condense and simplify the "core" repo structure
* improve consistency among the "core" repos
* ease the burden necessary to contribute functionality to the product
* reduce the level of overhead when attempting to publish new releases

The final structure for the "core" repos would look like:

```diff
1. cli
- compiler: moved -> server
- mock: moved -> server/worker
- pkg-executor: moved -> worker
- pkg-queue: moved -> server/worker
- pkg-runtime: moved -> worker
2. sdk-go
3. server
4. types
5. ui
6. worker
```

### CODE REVIEW

**Comment by kneal** [comment]
Definitely support this proposal! 😄  Would help with a lot of simplification across the project. I think another big call-out missing is how it would improve local development as well by allowing functional testing to be more approachable for certain components. 

The only additional thought I have is you could potentially make an argument for collapsing the `sdk-go` into the server repo. The only reason is if you're making an endpoint change on the router it should likely be reflected with the sdk's. However, we have a nice pattern for `sdk-<lang>` so I would vote to leave it separate.

**Comment by jbrockopp** [comment]
> Definitely support this proposal! 😄 Would help with a lot of simplification across the project.

Agreed 😃 

> I think another big call-out missing is how it would improve local development as well by allowing functional testing to be more approachable for certain components.

Yeah, I probably could have called it out as it's own individual problem 😞 

Instead, I listed it under [Problem 2 at the end](https://github.com/go-vela/community/blob/feature/proposals/2021/repo_structure/proposals/2021/08-25_repo-structure.md#problem-2) 👍 

I choose not to list it as its own individual problem due it being situational based depending on the change being made

> The only additional thought I have is you could potentially make an argument for collapsing the `sdk-go` into the server repo. The only reason is if you're making an endpoint change on the router it should likely be reflected with the sdk's. However, we have a nice pattern for `sdk-<lang>` so I would vote to leave it separate.

Agreed 😄 

That was also feedback I received from @JordanSussman when I shared the proposal with him 👍 

He even brought up the idea of automatically generating the SDK code from the swagger spec

This follows the pattern currently used in https://github.com/go-vela/sdk-python

Ultimately, I think that is something we could still look to do in the future but I don't want to make the proposal too big

So I think a separate proposal could (and should) be written for generating the SDKs and nesting them under the server

File: proposals/2021/08-25_repo-structure.md [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+The way the repos are setup today, there is a dependency tree that must be followed in order to properly release the product.
+
+The way the repos must be released is in groups that use the following order:
+
+1. [types](https://github.com/go-vela/types), [ui](https://github.com/go-vela/ui)
+2. [compiler](https://github.com/go-vela/compiler), [mock](https://github.com/go-vela/mock)
+3. [pkg-runtime](https://github.com/go-vela/pkg-runtime), [sdk-go](https://github.com/go-vela/sdk-go)
+4. [pkg-executor](https://github.com/go-vela/pkg-executor), [pkg-queue](https://github.com/go-vela/pkg-queue)
+5. [cli](https://github.com/go-vela/cli), [server](https://github.com/go-vela/server), [worker](https://github.com/go-vela/worker)
```
Comment (matt-fevold) [inline proposals/2021/08-25_repo-structure.md:167]: So, if I'm understanding this correctly with this proposal we'd end up eliminating steps 2 & 4 and step 3 would only be the `sdk-go`? Seems like quite the improvement 🎉

Comment (jbrockopp) [reply proposals/2021/08-25_repo-structure.md:167]: That is correct 😃

**Review by wass3r — APPROVED** [review]

**Review by KellyMerrick — APPROVED** [review]
lgtm

**Pull request merged**